### PR TITLE
Added missing validation in data ingress and refactoring

### PR DESCRIPTION
**Pull request by guilherme-brandao**
- After reviewing all the msgs, I saw some missing validations. I will address this in this PR
- Also saw some opportunities to do a small refactoring to avoid duplicate queries

### CODE REVIEW

File: x/emissions/keeper/keeper.go [hunk 1]
```diff
@@ -2011,6 +2008,10 @@ func (k *Keeper) GetTopic(ctx context.Context, topicId TopicId) (types.Topic, er
 
 // Sets a topic config on a topicId
 func (k *Keeper) SetTopic(ctx context.Context, topicId TopicId, topic types.Topic) error {
+	_, err := k.IncrementTopicId(ctx)
```
Comment (relyt29) [inline x/emissions/keeper/keeper.go:2011]: uhhhhh i think this is going to break a lot of unit tests that expect to be able to set topics

Comment (guilherme-brandao) [reply x/emissions/keeper/keeper.go]: Just reverted this change

**Review by kpeluso — DISMISSED** [review]

File: x/emissions/keeper/msgserver/msg_server_stake.go [hunk 2]
```diff
@@ -121,6 +124,15 @@ func (ms msgServer) CancelRemoveStake(ctx context.Context, msg *types.CancelRemo
 	if err := msg.Validate(); err != nil {
 		return nil, err
 	}
+
+	topicExists, err := ms.k.TopicExists(ctx, msg.TopicId)
+	if err != nil {
+		return nil, err
+	}
```
Comment (spooktheducks) [inline x/emissions/keeper/msgserver/msg_server_stake.go:131]: ditto

File: x/emissions/keeper/msgserver/msg_server_stake.go [hunk 3]
```diff
@@ -121,6 +124,15 @@ func (ms msgServer) CancelRemoveStake(ctx context.Context, msg *types.CancelRemo
 	if err := msg.Validate(); err != nil {
 		return nil, err
 	}
+
+	topicExists, err := ms.k.TopicExists(ctx, msg.TopicId)
```
Comment (spooktheducks) [inline x/emissions/keeper/msgserver/msg_server_stake.go:128]: Are there other fields we can proactively validate?  Be paranoid

File: x/emissions/keeper/msgserver/msg_server_stake.go [hunk 4]
```diff
@@ -68,6 +63,14 @@ func (ms msgServer) RemoveStake(ctx context.Context, msg *types.RemoveStakeReque
 		return nil, err
 	}
 
+	topicExists, err := ms.k.TopicExists(ctx, msg.TopicId)
```
Comment (spooktheducks) [inline x/emissions/keeper/msgserver/msg_server_stake.go:66]: Are there other fields we can proactively validate?  Be paranoid

File: x/emissions/keeper/msgserver/msg_server_stake.go [hunk 5]
```diff
@@ -144,12 +156,13 @@ func (ms msgServer) DelegateStake(ctx context.Context, msg *types.DelegateStakeR
 	if err := msg.Validate(); err != nil {
 		return nil, err
 	}
-	if msg.Amount.IsZero() {
-		return nil, types.ErrReceivedZeroAmount
-	}
 
-	if msg.Reputer == msg.Sender {
-		return nil, types.ErrCantSelfDelegate
+	topicExists, err := ms.k.TopicExists(ctx, msg.TopicId)
```
Comment (spooktheducks) [inline x/emissions/keeper/msgserver/msg_server_stake.go:160]: I'm happy that you're being super paranoid checking and re-checking this.  We can optimize later.  Thank you

File: x/emissions/keeper/msgserver/msg_server_stake.go [hunk 6]
```diff
@@ -144,12 +156,13 @@ func (ms msgServer) DelegateStake(ctx context.Context, msg *types.DelegateStakeR
...
+	if err != nil {
+		return nil, err
+	}
+	if !topicExists {
```
Comment (spooktheducks) [inline x/emissions/keeper/msgserver/msg_server_stake.go:164]: ditto from above about if/else

File: x/emissions/keeper/msgserver/msg_server_stake.go [hunk 7]
```diff
@@ -187,8 +200,13 @@ func (ms msgServer) RemoveDelegateStake(ctx context.Context, msg *types.RemoveDe
 	if err := msg.Validate(); err != nil {
 		return nil, err
 	}
-	if msg.Amount.LTE(cosmosMath.ZeroInt()) {
-		return nil, types.ErrInvalidValue
+
+	topicExists, err := ms.k.TopicExists(ctx, msg.TopicId)
```
Comment (spooktheducks) [inline x/emissions/keeper/msgserver/msg_server_stake.go:204]: Woof woof 💪

File: x/emissions/keeper/msgserver/msg_server_worker_payload.go [hunk 8]
```diff
@@ -17,11 +17,10 @@ import (
 func (ms msgServer) InsertWorkerPayload(ctx context.Context, msg *types.InsertWorkerPayloadRequest) (*types.InsertWorkerPayloadResponse, error) {
 	sdkCtx := sdk.UnwrapSDKContext(ctx)
 	blockHeight := sdkCtx.BlockHeight()
-	_, err := sdk.AccAddressFromBech32(msg.Sender)
-	if err != nil {
+	if err := ms.k.ValidateStringIsBech32(msg.Sender); err != nil {
```
Comment (spooktheducks) [inline x/emissions/keeper/msgserver/msg_server_worker_payload.go:20]: Scoping rules on `if` blocks are weird, so don't use them (see above)

File: x/emissions/keeper/msgserver/msg_server_worker_payload.go [hunk 9]
```diff
@@ -17,11 +17,10 @@ import (
...
 		return nil, err
 	}
-	err = checkInputLength(ctx, ms, msg)
+	err := checkInputLength(ctx, ms, msg)
```
Comment (spooktheducks) [inline x/emissions/keeper/msgserver/msg_server_worker_payload.go:23]: This is the way to do it 👍

File: x/emissions/module/rewards/rewards.go [hunk 10]
```diff
@@ -31,16 +32,12 @@ func EmitRewards(
...
-	if err != nil {
-		return errors.Wrapf(err, "failed to get module params")
-	}
-
 	// Sorted, active topics by weight descending. Still need skim top N to truly be the rewardable topics
 	sortedRewardableTopics := alloraMath.GetSortedElementsByDecWeightDesc(weights)
 	Logger(ctx).Debug(fmt.Sprintf("Rewardable topics: %v", sortedRewardableTopics))
 
 	if len(sortedRewardableTopics) == 0 {
+		fmt.Println("No rewardable topics found")
```
Comment (spooktheducks) [inline x/emissions/module/rewards/rewards.go:40]: Delete for prod

File: x/emissions/types/validations.go [hunk 11]
```diff
@@ -663,52 +663,61 @@ func (msg *RemoveRegistrationRequest) Validate() error {
...
+func stakeValidateHelper(addr []string, amount cosmosMath.Int, allowZeroAmount bool) error {
 	for _, ad := range addr {
 		_, err := sdk.AccAddressFromBech32(ad)
 		if err != nil {
 			return errors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid address (%s)", ad)
 		}
 	}
-	if amount.LT(cosmosMath.ZeroInt()) {
-		return errors.Wrapf(sdkerrors.ErrInvalidCoins, "invalid amount (%s)", amount.String())
+	if amount.IsNegative() {
```
Comment (spooktheducks) [inline x/emissions/types/validations.go:673]: Cheaper check, bump above for loop

File: x/emissions/types/validations.go [hunk 12]
```diff
@@ -663,52 +663,61 @@ func (msg *RemoveRegistrationRequest) Validate() error {
...
+		return errors.Wrapf(sdkerrors.ErrInvalidCoins, "amount must be non-negative: %s", amount.String())
 	}
+
+	if !allowZeroAmount && amount.IsZero() {
```
Comment (spooktheducks) [inline x/emissions/types/validations.go:677]: Same

File: x/emissions/types/validations.go [hunk 13]
```diff
@@ -663,52 +663,61 @@ func (msg *RemoveRegistrationRequest) Validate() error {
...
 // validate that a remove stake request follows the expected format
 func (msg *RemoveStakeRequest) Validate() error {
-	return stakeValidateHelper([]string{msg.Sender}, msg.Amount)
+	return stakeValidateHelper([]string{msg.Sender}, msg.Amount, false)
 }
 
 // validate that a delegate stake request follows the expected format
 func (msg *DelegateStakeRequest) Validate() error {
-	return stakeValidateHelper([]string{msg.Sender, msg.Reputer}, msg.Amount)
+	if msg.Reputer == msg.Sender {
```
Comment (spooktheducks) [inline x/emissions/types/validations.go:696]: Perfect, keep up this way of thinking

File: x/emissions/types/validations.go [hunk 14]
```diff
@@ -663,52 +663,61 @@ func (msg *RemoveRegistrationRequest) Validate() error {
...
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "cannot self delegate")
+	}
+
+	return stakeValidateHelper([]string{msg.Sender, msg.Reputer}, msg.Amount, false)
 }
 
 // validate that a remove delegate stake request follows the expected format
 func (msg *RemoveDelegateStakeRequest) Validate() error {
-	return stakeValidateHelper([]string{msg.Sender, msg.Reputer}, msg.Amount)
+	return stakeValidateHelper([]string{msg.Sender, msg.Reputer}, msg.Amount, false)
```
Comment (spooktheducks) [inline x/emissions/types/validations.go:705]: Can you leave some comments on this and the following methods explaining the boolean?  i.e. Why do these various request types allow or disallow zero amounts

File: x/emissions/types/validations.go [hunk 15]
```diff
@@ -663,52 +663,61 @@ func (msg *RemoveRegistrationRequest) Validate() error {
 }
 
 // helper for stake validation functions
-func stakeValidateHelper(addr []string, amount cosmosMath.Int) error {
+func stakeValidateHelper(addr []string, amount cosmosMath.Int, allowZeroAmount bool) error {
```
Comment (spooktheducks) [inline x/emissions/types/validations.go:666]: Better comment here would be useful as this helper is used several times below.  Might supersede my suggestion to comment each function below if it's written well

**Review by spooktheducks — CHANGES_REQUESTED** [review]

File: x/emissions/keeper/msgserver/msg_server_reputer_payload.go [hunk 16]
```diff
@@ -43,7 +42,7 @@ func (ms msgServer) InsertReputerPayload(ctx context.Context, msg *types.InsertR
 	}
 	// Returns an error if unfulfilled worker nonce exists
 	if workerNonceUnfulfilled {
-		return nil, types.ErrNonceStillUnfulfilled
+		return nil, errorsmod.Wrapf(types.ErrNonceStillUnfulfilled, "worker nonce")
```
[the remark this answers is not in the document]
Comment (relyt29) [reply x/emissions/keeper/msgserver/msg_server_reputer_payload.go:45]: it's a cosmos-sdk library that builds on top of the standard go library errors package - you still get the standard stack trace with line numbers but the "Wrap/Wrapf" makes the error message have this additional string in it when the error message at the bottom of the stack trace is printed

File: x/emissions/keeper/msgserver/msg_server_reputer_payload.go [hunk 17]
```diff
@@ -11,17 +11,16 @@ import (
 
 // A tx function that accepts a individual loss and possibly returns an error
 func (ms msgServer) InsertReputerPayload(ctx context.Context, msg *types.InsertReputerPayloadRequest) (*types.InsertReputerPayloadResponse, error) {
-	_, err := sdk.AccAddressFromBech32(msg.Sender)
-	if err != nil {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	blockHeight := sdkCtx.BlockHeight()
+	if err := ms.k.ValidateStringIsBech32(msg.Sender); err != nil {
```
[the remark this answers is not in the document]
Comment (kpeluso) [reply x/emissions/keeper/msgserver/msg_server_reputer_payload.go:16]: will need a ticket to universally fix this + new linter

[the remark this answers is not in the document]
Comment (kpeluso) [reply x/emissions/keeper/msgserver/msg_server_reputer_payload.go:16]: https://linear.app/alloralabs/issue/PROTO-2494/parse-all-parts-in-code-then-add-linterpre-commit-hook-to-prevent

**Comment by relyt29** [comment]
PLEASE REBASE THIS BRANCH ONTO DEV AGAIN, AS DEV HAS BEEN PUSHED TO A NEW HEAD THAT REMOVES A BAD COMMIT
```
git fetch 
git rebase -i origin/dev
```
DO NOT UPDATE IN GITHUB UI, YOU WILL REINTRODUCE THE BAD COMMIT IN YOUR MERGE COMMIT - I REPEAT DO NOT MERGE

@guilherme-brandao

**Review by kpeluso — APPROVED** [review]

File: x/emissions/keeper/keeper.go [hunk 18]
```diff
@@ -2076,6 +2073,7 @@ func (k *Keeper) GetStakeRemovalForReputerAndTopicId(
 	if err != nil {
 		return types.StakeRemovalInfo{}, false, errorsmod.Wrap(err, "error iterating over stake removals by actor")
 	}
+	defer iter.Close()
```
Comment (RedBird96) [inline x/emissions/keeper/keeper.go:2076]: good catch

**Review by RedBird96 — APPROVED** [review]

**Comment by relyt29** [comment]
can you update the description of this PR to a bit more descriptive of what it does / tests documentation changelog etc

**Review by spooktheducks — APPROVED** [review]

**Comment by guilherme-brandao** [comment]
> can you update the description of this PR to a bit more descriptive of what it does / tests documentation changelog etc

Merged now to avoid conflicts. I will be updating the description soon

**Pull request merged**